src/plugins/mpath.c

```diff
@@ -156,360 +184,295 @@ gboolean bd_mpath_is_tech_avail (BDMpathTech tech, guint64 mode, GError **error)
  * Tech category: %BD_MPATH_TECH_BASE-%BD_MPATH_TECH_MODE_MODIFY
  */
 gboolean bd_mpath_flush_mpaths (GError **error) {
-    const gchar *argv[3] = {"multipath", "-F", NULL};
-    gboolean success = FALSE;
-    gchar *output = NULL;
+    struct dmmp_mpath **mps = NULL;
+    struct dmmp_context *ctx = NULL;
+    uint32_t mp_count = 0;
+    uint32_t i = 0;
+    gboolean rc = FALSE;
+    const char *name = NULL;
+    int dmmp_rc = DMMP_OK;
+    char err_msg_buff[_ERR_MSG_BUFF_SIZE];
 
-    if (!check_deps (&avail_deps, DEPS_MPATH_MASK, deps, DEPS_LAST, &deps_check_lock, error))
-        return FALSE;
+    assert (error != NULL);
 
-    /* try to flush the device maps */
-    success = bd_utils_exec_and_report_error (argv, NULL, error);
-    if (!success)
-        return FALSE;
+    *error = NULL;
 
-    /* list devices (there should be none) */
-    argv[1] = "-ll";
-    success = bd_utils_exec_and_capture_output (argv, NULL, &output, error);
-    if (success && output && (g_strcmp0 (output, "") != 0)) {
-        g_set_error (error, BD_MPATH_ERROR, BD_MPATH_ERROR_FLUSH,
-                     "Some device cannot be flushed: %s", output);
-        g_free (output);
+    if (geteuid () != 0) {
+        g_set_error (error, BD_MPATH_ERROR, BD_MPATH_ERROR_NOT_ROOT,
+                     "Not running as root, cannot flush mpaths");
         return FALSE;
     }
 
-    g_free (output);
-    return TRUE;
-}
-
-static gchar* get_device_name (const gchar *major_minor, GError **error) {
-    gchar *path = NULL;
-    gchar *link = NULL;
-    gchar *ret = NULL;
-
-    path = g_strdup_printf ("/dev/block/%s", major_minor);
-    link = g_file_read_link (path, error);
-    g_free (path);
-    if (!link) {
-        g_prefix_error (error, "Failed to determine device name for '%s'",
-                        major_minor);
-        return NULL;
-    }
-
-    /* 'link' should be something like "../sda" */
-    /* get the last '/' */
-    ret = strrchr (link, '/');
-    if (!ret) {
-        g_set_error (error, BD_MPATH_ERROR, BD_MPATH_ERROR_INVAL,
-                     "Failed to determine device name for '%s'",
-                     major_minor);
-        g_free (link);
-        return NULL;
+    _dmmp_ctx_setup (ctx, err_msg_buff);
+    if (ctx == NULL) {
+        g_set_error (error, BD_MPATH_ERROR, BD_MPATH_ERROR_NO_MEMORY,
+                     "No enough memory to create dmmp context");
+        goto out;
```

Comment:
Please avoid the ``goto`` statement. I know it make things more complicated, but we have always been sacrificing that for avoiding ``goto``s in this library and this doesn't look like any special case that should end such efforts.

Reply:
Sure. Will do.

Reply:
Just FYI: https://peerj.com/preprints/826v1/ 
```
We conclude that developers limit themselves to using 
goto appropriately in most cases, and not in an unrestricted 
manner like Dijkstra feared, thus suggesting that goto does 
not appear to be harmful in practice.
```